crates/ruff_python_parser/src/lexer.rs

```diff
@@ -508,6 +514,138 @@ impl<'source> Lexer<'source> {
         }
     }
 
+    /// Lex a f-string start token.
+    fn lex_fstring_start(&mut self, quote: char, triple_quoted: bool, is_raw_string: bool) -> Tok {
+        let quote_size = if triple_quoted {
+            StringQuoteSize::Triple
+        } else {
+            StringQuoteSize::Single
+        };
+        // SAFETY: Safe because `quote` is either `'` or `"`
+        let quote_char = StringQuoteChar::try_from(quote).unwrap();
+        self.fstring_stack
+            .push(FStringContext::new(quote_char, quote_size, is_raw_string));
+        Tok::FStringStart
+    }
+
+    fn try_lex_fstring_middle(&mut self) -> Result<Option<Tok>, LexicalError> {
+        let mut in_named_unicode = false;
+
+        // SAFETY: Safe because the function is only called when `self.fstring_stack` is not empty.
+        let context = self.fstring_stack.last().unwrap();
```

Comment:
Could we pass the fstring context instead of using an `unwrap` here? It isn't just safer but also guarantees that we don't have a panic branch in release builds.

Reply:
Not really. The `self.try_lex_fstring` functions borrow `self` as mutable but then we pass in an immutable borrow of `FStringContext` which belongs to `self`:

```
error[E0502]: cannot borrow `*self` as mutable because it is also borrowed as immutable
   --> crates/ruff_python_parser/src/lexer.rs:730:36
    |
727 |         if let Some(fstring_context) = self.fstring_stack.last() {
    |                                        ------------------------- immutable borrow occurs here
...
730 |                 if let Some(tok) = self.try_lex_fstring_middle(fstring_context)? {
    |                                    ^^^^^----------------------^^^^^^^^^^^^^^^^^
    |                                    |    |
    |                                    |    immutable borrow later used by call
    |                                    mutable borrow occurs here
```